### PR TITLE
Reimplement `enso_project` as a proper builtin

### LINKED ISSUE #5050: `enso_project` does not appear in Component Browser suggestions
This task is automatically imported from the old Task Issue Board and it was originally created by Radosław Waśko.
Original issue is [here](https://www.pivotaltracker.com/story/show/184380092).
<hr />

As in the title, the `enso_project` special method is not discoverable as it does not appear in suggestions.

## Comments:
 (Radosław Waśko - Feb 2, 2023)
<hr />

### PR DESCRIPTION
**Pull request by Akirathan**
Fixes #5050

### Pull Request Description

Remove the magical code generation of `enso_project` method from codegen phase and reimplement it as a proper builtin method.

The old behavior of `enso_project` was special, and violated the language semantics (regarding the `self` argument):
- It was implicitly declared in every module, so it could be called without a self argument.
- It can be called with explicit module as self argument, e.g. `Base.enso_project`, or `Visualizations.enso_project`.

Let's avoid implicit methods on modules and let's be explicit. Let's reimplement the `enso_project` as a builtin method. To comply with the language semantics, we will have to change the signature a bit:
- `enso_project` is a static method in the `Standard.Base.Meta.Enso_Project` module.
- It takes an optional `project` argument (instead of taking it as an explicit self argument).

Having the `enso_project` defined as a (shadowed) builtin method, we will automatically have suggestions created for it.

### Important Notes
- Truffle nodes are no longer generated in codegen phase for the `enso_project` method. It is a standard builtin now.
  - The minimal import to use `enso_project` is now `from Standard.Base.Meta.Enso_Project import enso_project`.
- Tested implicitly by `org.enso.compiler.ExecCompilerTest#testInvalidEnsoProjectRef`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follow the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

**Comment by 4e6** [comment]
Please also add tests checking that 
- the `enso_project` method works properly in interpreter mode. See the [RuntimeServerTest](https://github.com/enso-org/enso/blob/2d168308afc44d27a90e12b1ca91ba04890f6552/engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala) as an example. Notice that it uses `micro-distribution` instead of standard `distribution` library, and you may need to extend it with `enso_project` method as well
- the engine generates proper `enso_project` suggestion in [RuntimeStdlibTest](https://github.com/enso-org/enso/blob/2d168308afc44d27a90e12b1ca91ba04890f6552/engine/runtime-with-instruments/src/test/scala/org/enso/interpreter/test/instrument/RuntimeStdlibTest.scala)

File: CHANGELOG.md [hunk 1]
```diff
@@ -801,8 +802,11 @@
 [6151]: https://github.com/enso-org/enso/pull/6151
 [6171]: https://github.com/enso-org/enso/pull/6171
 [5998]: https://github.com/enso-org/enso/pull/5998
+
 [6335]: https://github.com/enso-org/enso/pull/6335
 
+[6352]: https://github.com/enso-org/enso/pull/6352
```
Comment (radeusgd) [inline CHANGELOG.md]: Are these newlines expected?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso [hunk 2]
```diff
@@ -48,3 +50,28 @@ type Project_Description
              enso_project.namespace
     namespace : Text
     namespace self = self.prim_config.namespace
+
+## Returns the Enso project description for the given module.
+
+   Arguments:
+   - module: An optional module for which the Enso project description
+     will be fetched.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso]: What is the behaviour if the module is not provided?

Ah, I see in the example below - it's the 'current project'. Maybe let's note this in the `Arguments` too.

File: distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso [hunk 3]
```diff
@@ -18,11 +18,11 @@ type Example_Error_Type
 
 ## The standard library data directory.
 data_dir : File
-data_dir = enso_project.data
+data_dir = (enso_project Standard.Examples).data
```
Comment (radeusgd) [inline distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: Why is this change needed?

I assume the default should work OK here.

File: CHANGELOG.md (hunk 1 above)
Comment (Akirathan) [reply CHANGELOG.md]: No, seems like automatic formatter. Fixed.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso (hunk 2 above)
Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso]: Fixed in b9c3e744de3add8c0cd982e22266ecd6233b4538.

File: distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso (hunk 3 above)
Comment (Akirathan) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: The problem is once the `data_dir` method is called from a module outside `Examples` project. For example, there is `Examples_Tests` project, that uses `Examples.data_dir`. So, e.g., for the following call hierarchy:
```
Examples_Tests.my_test --> Examples.data_dir --> enso_project
```
It will return `Project_Description("Examples_Tests")`. I think that is better than the old behavior. What do you think? Besides, it forces us exactly on these places to explicitly specify the module for which we want the `Project_Description`.

Comment (radeusgd) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: Hmmm. So this is a change of semantics from what was there before?

I'm slightly worried about this change. It does align with how the new syntax work, but from user perspective - now it's basically never a good idea to call `enso_project` without specifying the argument explicitly - because as you've shown - _now_ if I don't specify the argument, the result of such a method will depend on _where it is called from_. I'm really not sure if this method's result should depend on the runtime call-site and not where 'statically' it is in the code.

I expect at some point when people start using `enso_project` in places other than directly in the GUI, it may very easily bite us.

Maybe we should keep `enso_project` without arguments and note that it returns the _currently opened project regardless of where it is called in_? (i.e. in the IDE it would always return the current project, even if called in some other module - i.e. it would work like the current default behaviour)

And then we could have a separate helper `find_project_for_module` taking the argument?

I feel like having both these functionalities in a single method may lead to mistakes.

Comment (Akirathan) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: OK. Sounds reasonable. Let me try to see if I can tweak it such that `enso_project` without any arguments will resolve to the current project from the "statically" perspective, not from the project from where it was called.

Comment (JaroslavTulach) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: > `enso_project` without any arguments will resolve to the current project from the "statically" perspective

Get the call stack. Walk-up one frame. Find caller. Find its `RootNode`. If it is `EnsoRootNode` call its `getModuleScope()`. Cache the value (per `EnsoContext`) and make it compile time constant (at least for a finite number of contexts).

> this is a change of semantics from what was there before?

It is preferable to avoid changes of the semantics.

Comment (Akirathan) [reply distribution/lib/Standard/Examples/0.0.0-dev/src/Main.enso]: > Get the call stack. Walk-up one frame. Find caller. Find its RootNode. If it is EnsoRootNode call its getModuleScope().

Done in 005fb3807bd8590b35baea2c7e7e72f48f1c6f3e

> Cache the value (per EnsoContext) and make it compile time constant (at least for a finite number of contexts).

What do you mean exactly by caching the value per `EnsoContext`? The Project Description depends on the location of the caller in the source code, not on the context. In one `EnsoContext`, there might be multiple callers from different modules. Am I misunderstanding your proposition?

**Comment by Akirathan** [comment]
Tested with the IDE:
image

Moreover, there is now a suggestion entry in the CB:
image

It seems like the issue is resolved.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+   */
+  @Specialization(guards = "isNothing(interop, nothing)")
+  @TruffleBoundary
+  public Object getCurrentProjectDescr(
+      Object nothing, @CachedLibrary(limit = "5") InteropLibrary interop) {
+    var ctx = EnsoContext.get(this);
+    // Find the caller of `enso_project`, i.e., of this node, and find in which package
+    // it is located. The first frame is skipped, because it is always `Enso_Project.enso_project`,
+    // i.e., the first frame is always call of this specialization.
+    Optional<Package<TruffleFile>> pkgOpt =
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: Opinion: Java code is better without `Optional`. Easier to use `null`. But I see `null` has special meaning for the iteration, so I am able to live with it.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
+import org.enso.pkg.Package;
+
+@BuiltinMethod(
+    type = "Project_Description",
+    name = "enso_project_builtin",
+    description = "Returns the project description of the project given as the argument")
+public abstract class EnsoProjectNode extends Node {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java:27]: This node is inserted into AST of the module that uses the builtin. E.g. there are different instances of this node in different sources and source locations.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+   * call of {@code enso_project}) - the caller of {@code enso_project}, and finds in which package
+   * the caller is located.
+   *
+   * @param nothing Nothing, or interop null.
+   */
+  @Specialization(guards = "isNothing(interop, nothing)")
+  @TruffleBoundary
+  public Object getCurrentProjectDescr(
+      Object nothing, @CachedLibrary(limit = "5") InteropLibrary interop) {
+    var ctx = EnsoContext.get(this);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java:57]: TCR: The result of this method is always the same (as the node is inserted at single source location)  for every `ctx` - cache it and avoid `@TruffleBoundary` on subsequent invocations.

File: engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala [hunk 7]
```diff
@@ -370,8 +370,8 @@ final class SerializationManager(
           module.setLoadedFromCache(true)
           logger.log(
             debugLogLevel,
-            "Restored IR from cache for module [{0}] at stage [{1}].",
-            Array(module.getName, loadedCache.compilationStage())
+            () =>
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: TCR: I find the previous version better. Single text with `{0}` placeholders and an array with values to fill in. Such a style allows for example [structured logging](https://bits.netbeans.org/dev/javadoc/org-openide-util/org/openide/util/doc-files/logging.html). What's the justification for this change?

File: test/Tests/src/Semantic/Meta_Location_Spec.enso [hunk 8]
```diff
@@ -25,6 +25,9 @@ spec = Test.group "Meta-Value Inspection" <|
 
     Test.specify "should allow access to package names" <|
         enso_project.name.should_equal 'Tests'
-        Base.enso_project.name.should_equal 'Base'
+        (enso_project Base).name.should_equal 'Base'
+
+    Test.specify "should allow to fetch enso project description from a module" <|
+        (enso_project Standard.Base.Data.Vector).name.should_equal "Base"
```
Comment (JaroslavTulach) [inline test/Tests/src/Semantic/Meta_Location_Spec.enso:31]: Acceptable changes in the API. It is important line 27 remains unchanged.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        this);
+  }
+
+  private DataflowError notInModuleError(EnsoContext ctx) {
+    return DataflowError.withoutTrace(
+        ctx.getBuiltins().error().makeModuleNotInPackageError(), this);
+  }
+
+  boolean isNothing(InteropLibrary interop, Object object) {
+    return interop.isNull(object);
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: TCA: It'd be more lightweight to check just `ctx.getBuiltins().nothing() == object`.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso [hunk 10]
```diff
@@ -42,6 +44,8 @@ export project.Errors
 export project.IO
 export project.Math
 export project.Meta
+export project.Meta.Enso_Project
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso]: Is it important to export `Enso_Project` symbol? Isn't it enough to have just `enso_project`?

**Review by JaroslavTulach — APPROVED** [review]
I am glad to see such an amount of removed _"special purpose"_ code. I left few comments:
- TCR - technical change required - shall in my opinion be done
- TCA - technical change advised - would be nice to be done
- the rest are just opinions

File: engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala (hunk 7 above)
Comment (Akirathan) [reply engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: Because the previous result of the logging of Array is, e.g., like this:
```log
[debug] [2023-04-18T10:06:25.573Z] [enso.org.enso.compiler.SerializationManager] Restored IR from cache for module [[Ljava.lang.Object;@2ed09493] at stage [{1}]
```
Probably the `logger.log` is not able to cope with Scala's `Array`? Anyway, I always use lambda functions for loggers. I find it most straightforward and if I understand that correctly, it should also have a lower overhead if the logger is disabled, because you don't have to create any objects.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java (hunk 6 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java:57]: Add two `@CompilationFinal` fields into the `EnsoProjectNode`:
```
@CompilationFinal
private Reference<EnsoContext> previousCtxt;
@CompilationFinal
private Reference<Atom> previousProject;
```
and speculate on `EnsoContext` being always the same. Then the result shall also be the same.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java (hunk 9 above)
Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: Is this surely valid? I'm not that well versed in polyglot, but `ctx.getBuiltins().nothing()` is the Enso Nothing singleton. What about 'nulls' from other languages?

If I do:
```
foreign js my_method = """
    return null;

enso_project my_method
```
then will this work correctly? I'm not sure when is the null value converted to the Enso value?

(I know this example is weird and who'd create a `null` in JS to pass it over to `enso_project` - but it can be done so if it happens we want to be sure it will not end up with a weird unexpected behaviour)

File: engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala (hunk 7 above)
Comment (Akirathan) [reply engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: Fixed in 45d9301e899f896752b28db65ce84867453afd4a. Seems like the required change was to explicitly cast `Array` to `Array[Object]`.

Comment (Akirathan) [reply engine/runtime/src/main/scala/org/enso/compiler/SerializationManager.scala]: Fixed in 45d9301e899f896752b28db65ce84867453afd4a without using lambda.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java (hunk 6 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java:57]: Done in 9355e70d8da2721b6705289ef31e72e3cc88dcde

**Comment by somebody1234** [comment]
thonk

**Comment by somebody1234** [comment]
might be missing a merge commit

**Comment by Akirathan** [comment]
omg, seems like I somehow messed the merge commit .. sorry guys for spam, I will try to fix that

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java (hunk 9 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: @radeusgd is correct in this case. `InteropLibrary.isNull` is more general and, therefore, should deal even with these cases when called with a polyglot null.

Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: There is a cost of `InteropLibrary` (polymorphic, heavyweight, requiring too much `Node` instances). The only reason for `Nothing` is to support a _"default value"_. There is no point in supporting JavaScript `null` instead. Any value different from `Nothing|Module` may easily yield an error.

If we really want to support such foreign values, then we should have a specific `IsNullNode` or `IsNothingNode` and use it. It would specialize for `Nothing`, but could handle interop values as well, when they appear. Similar to #6299 idea.

Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EnsoProjectNode.java]: `InteropLibrary` overhead removed in e2a440d7f1b4bed284b0b5b45885524823044d9e. @radeusgd Note that if the wrong argument is passed, Unsupported_Arguments error is returned.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso (hunk 10 above)
Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso]: I am afraid so, this is the only way how to export a static method from a module. See for example, https://github.com/enso-org/enso/tree/develop/engine/runtime/src/test/resources/Test_Multiple_Exports.

Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso]: Hmm, I have just tried removing the import/export of the module, and everything seems to work, even on the IDE side. Commit: 1843fdc5dda5ba9c8b8e5c49321f330f73ec1aa3

File: build.sbt [hunk 11]
```diff
@@ -1074,6 +1074,7 @@ lazy val `polyglot-api` = project
 
 lazy val `language-server` = (project in file("engine/language-server"))
   .settings(
+    commands += WithDebugCommand.withDebug,
```
Comment (Akirathan) [inline build.sbt:1078]: This is useful for debugging tests in `language-server` project with `withDebug` command.

File: engine/runtime/src/test/java/org/enso/compiler/ExecCompilerTest.java [hunk 12]
```diff
@@ -68,8 +68,12 @@ public void testHalfAssignment() throws Exception {
 
   @Test
   public void testInvalidEnsoProjectRef() throws Exception {
-    var module = ctx.eval("enso", """
+    var module =
+        ctx.eval(
+            "enso",
+            """
     from Standard.Base.Errors.Common import all
+    from Standard.Base.Meta.Enso_Project import enso_project
```
Comment (Akirathan) [inline engine/runtime/src/test/java/org/enso/compiler/ExecCompilerTest.java:76]: We have to explicitly import `enso_project` here, as `enso_project` is no longer a special generated method on every module. Note that `enso_project` is exported by default with `from Standard.Base import all`.

**Review by MichaelMauderer — APPROVED** [review]

**Review by radeusgd — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso [hunk 13]
```diff
@@ -47,3 +49,29 @@ type Project_Description
...
+
+   Arguments:
+   - module: An optional module for which the Enso project description
+     will be fetched. If `Nothing`, the current project description will
+     be fetched.
+
+   > Example
+     Get the project description for the current project.
+
+         enso_project
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso]: Can we clarify what does `current project` mean?

If I have:
```
Project_1/
    src/
        Main.enso:
             from Standard.Base import all
             foo = IO.println enso_project
Project_2/
    src/
        Main.enso:
             import local.Project_1
             main = Project_1.foo
```

will I get `Project_1` or `Project_2`?

**Review by GregoryTravis — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso (hunk 13 above)
Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso]: In your example it would return `Project_2`. Docs updated in 6a5f147ce8c1428712b20fde55a9ae6cd40e54e8

Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Meta/Enso_Project.enso]: Awesome.

**Review by jdunkerley — APPROVED** [review]
As the principal API still works am happy with this.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso [hunk 14]
```diff
@@ -42,6 +43,7 @@ export project.Errors
 export project.IO
 export project.Math
 export project.Meta
+from project.Meta.Enso_Project export enso_project
```
Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso:46]: We should probably add `enso_project` to `micro-distribution`

**Review by hubertp — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso (hunk 14 above)
Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso:46]: Done in ef5deaf3c6ac5f48c3e5d77c5c794a28e4594880

Comment (Akirathan) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Main.enso:46]: Even though the module layout is not the same as in the standard distribution, but that should not be a problem, right?

**Comment by Akirathan** [comment]
Cannot reproduce 
```
org.graalvm.polyglot.PolyglotException: java.lang.AssertionError: assertion failed: exported symbol `enso_project` needs to be registered first in the module
```
(https://github.com/enso-org/enso/actions/runs/4839795157/jobs/8625338504#step:10:765) locally. Ensure "CI: Clean build required" label is set and restarting the jobs.

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Reimplement `enso_project` as a proper builtin · enso-org/enso@615847a</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by Akirathan** [comment]
Forcing clean build with [clean action](https://github.com/enso-org/ci/actions/runs/4859867515). Seems like `CI: Clean build require` does not work as expected?
